Mathlib/NumberTheory/SelbergSieve.lean

```diff
@@ -222,4 +225,200 @@ theorem siftedSum_le_mainSum_errSum_of_upperMoebius (muPlus : ℕ → ℝ) (h :
     rw [← abs_mul]
     exact le_abs_self (muPlus d * s.rem d)
 
+
+
+section LambdaSquared
+/-- We consider a special class of upper bound sieves called the Λ² sieve. This class is
+  parameterised by a sequence of real numbers. We will later choose a set of weights that minimises
+  the main term, under a constraint that lets us control the error term. -/
+def lambdaSquared (weights : ℕ → ℝ) : ℕ → ℝ := fun d =>
+  ∑ d1 ∈ d.divisors, ∑ d2 ∈ d.divisors, if d = Nat.lcm d1 d2 then weights d1 * weights d2 else 0
+
+theorem lambdaSquared_eq_zero_of_not_le_height (w : ℕ → ℝ) (height : ℝ)
+    (hw : ∀ d : ℕ, ¬d ^ 2 ≤ height → w d = 0) (d : ℕ) (hd : ¬d ≤ height) :
+    lambdaSquared w d = 0 := by
+  apply sum_eq_zero; intro d1 hd1; apply sum_eq_zero; intro d2 hd2
```

Comment:
```suggestion
  refine sum_eq_zero fun d1 hd1 ↦ sum_eq_zero fun d2 hd2 ↦ ?_
```
One can fold the `apply; intro` indiom into one `refine`. I'll add this as suggestions, but ultimately, it is a matter of taste which one to prefer. One advantage of the `refine` version is that it avoids the use of `;`/requires fewer lines.